### ISSUE TITLE
dns-addon shows abnormally high restart rates

### ISSUE THREAD

**Issue by celskeggs** [sipb/homeworld#120]
We're up to 18 restarts within the first hour of DC5 having the dns-addon deployed. That's bad. What's going on?

**Comment by celskeggs** [comment]
Here's a snippet of the logs, in case anyone has time to dig deeper into this:

	Events:
	  Type     Reason                 Age                 From                     Message
	  ----     ------                 ----                ----                     -------
	  Normal   Scheduled              49m                 default-scheduler        Successfully assigned kube-dns-v20-wm79q to avocado-burger
	  Normal   SuccessfulMountVolume  49m                 kubelet, avocado-burger  MountVolume.SetUp succeeded for volume "default-token-7kqhs"
	  Normal   Pulling                49m                 kubelet, avocado-burger  pulling image "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.5"
	  Normal   Pulled                 49m                 kubelet, avocado-burger  Successfully pulled image "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.5"
	  Normal   Pulling                49m                 kubelet, avocado-burger  pulling image "gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.5"
	  Normal   Pulled                 49m                 kubelet, avocado-burger  Successfully pulled image "gcr.io/google_containers/k8s-dns-dnsmasq-nanny-amd64:1.14.5"
	  Normal   Pulling                49m                 kubelet, avocado-burger  pulling image "gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.5"
	  Normal   Pulled                 48m                 kubelet, avocado-burger  Successfully pulled image "gcr.io/google_containers/k8s-dns-sidecar-amd64:1.14.5"
	  Normal   Started                48m                 kubelet, avocado-burger  Started with rkt id dc2d59d5
	  Normal   Created                48m                 kubelet, avocado-burger  Created with rkt id dc2d59d5
	  Normal   Created                48m                 kubelet, avocado-burger  Created with rkt id dc2d59d5
	  Normal   Started                48m                 kubelet, avocado-burger  Started with rkt id dc2d59d5
	  Normal   Started                48m                 kubelet, avocado-burger  Started with rkt id dc2d59d5
	  Normal   Created                48m                 kubelet, avocado-burger  Created with rkt id dc2d59d5
	  Warning  Unhealthy              48m                 kubelet, avocado-burger  Readiness probe failed: Get http://172.18.5.2:8081/readiness: dial tcp 172.18.5.2:8081: getsockopt: connection refused
	  Warning  Unhealthy              47m (x4 over 47m)   kubelet, avocado-burger  Liveness probe failed: Get http://172.18.5.2:10054/healthcheck/dnsmasq: dial tcp 172.18.5.2:10054: getsockopt: connection refused
	  Warning  Unhealthy              28m (x30 over 47m)  kubelet, avocado-burger  Liveness probe failed: Get http://172.18.5.2:10054/metrics: dial tcp 172.18.5.2:10054: getsockopt: connection refused
	  Warning  Unhealthy              14m (x13 over 46m)  kubelet, avocado-burger  Liveness probe failed: Get http://172.18.5.5:10054/healthcheck/kubedns: dial tcp 172.18.5.5:10054: getsockopt: connection refused
	  Normal   Pulled                 9m (x29 over 47m)   kubelet, avocado-burger  Container image "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.5" already present on machine
	  Warning  Unhealthy              4m (x51 over 47m)   kubelet, avocado-burger  Liveness probe failed: Get http://172.18.5.2:10054/healthcheck/kubedns: dial tcp 172.18.5.2:10054: getsockopt: connection refused

**Comment by celskeggs** [comment]
This appears to be a problem with the `:10054` address in the container being inaccessible, which *should* be provided by the sidecar, but doesn't appear to be. In fact, none of the expected logging from the sidecar container (including startup version messages) appears to ever show up.

I also tried to switch the dns-addon to run outside the kvm stage1, in case that was the problem, but now it just has trouble contacting the kube-proxy services (and flannel endpoints on other hosts, for that matter).

**Comment by celskeggs** [comment]
After fixing #257, it seems that this actually works correctly iff it is run under the coreos stage1 instead of the kvm stage1.

This means that this is a bug with the kvm stage1.

**Comment by celskeggs** [comment]
Useful debugging tip: you can look under /run/systemd/system/ for the dynamically-generated systemd services to launch rkt.

```
root@FRENCH-TOAST:~# cat /run/systemd/system/k8s_d1a8aabc-d958-4c70-879b-6b82061f68ad.service
[Service]
ExecStart=/usr/bin/rkt --insecure-options=image,ondisk,none --local-config=/etc/rkt --system-config=/usr/lib/rkt --dir=/var/lib/rkt run-prepared --net=rkt.kubernetes.io --dns=18.70.0.160 --dns=18.71.0.151 --dns=18.72.0.3 --dns-search=MIT.EDU --dns-opt=ndots:5 --hostname=kube-dns-v20-j2cj6 d1a8aabc-d958-4c70-879b-6b82061f68ad
ExecStopPost=/usr/bin/touch /var/lib/kubelet/pods/48dc0a5e-1433-11e8-893d-525400ed9b5a/finished-d1a8aabc-d958-4c70-879b-6b82061f68ad
KillMode=mixed
TimeoutStopSec=30s
SyslogIdentifier=kube-dns-v20

[Unit]
Description=kube-dns-v20-j2cj6

[X-Kubernetes]
PodUID=48dc0a5e-1433-11e8-893d-525400ed9b5a
PodName=kube-dns-v20-j2cj6
PodNamespace=kube-system
PodHostNetwork=false
PodNetworkNamespace=
```

**Comment by celskeggs** [comment]
I ran kubelet with -v=4 so that I could extract the generated pod manifests. [Here's the manifest](https://gist.github.com/celskeggs/e759175b3c16d4f9cc0d38fd2afe7823), but I can't find anything wrong with it, unless perhaps there's a problem with not enough memory or CPU being allocated to the sidecar container? But I suspect I'd get an explicit error in that case...

**Comment by celskeggs** [comment]
The problem is not anything along the lines of "the third container won't start", because reordering the containers in the pod so that sidecar goes first doesn't change anything. Sidecar still doesn't launch.

**Comment by celskeggs** [comment]
I'm almost 100% certain that this is a resource-constraint issue. After manually launching the pod with rkt (by rebuilding the generated manifest to run outside of kubernetes), it seems that sidecar does eventually start... but several minutes later, at which point the pod has already been marked as failed.

This is because it's limited to 10m CPU, aka `CPUQuota=1%`. This makes it go *incredibly slow*, especially on this old platform, and the timeouts are missed.

I've deployed a new version of the spec that doesn't add resource constraints to the sidecar, and it seems to be stable.

I can't believe I ruled out the possibility of resources being the problem, simply because it had the same amount of _memory_ as another working container. I didn't realize that low CPU could just slow down the startup time immensely...
